sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CryptographicKeyShouldNotBeTooShort.cs

```diff
@@ -34,42 +35,243 @@ namespace SonarAnalyzer.Rules.CSharp
+                            break;
+                        case "GenerateKey":
+                            CheckSystemSecurityEllipticCurve(containingType.Value, invocation, invocation.ArgumentList, c);
+                            break;
+                        case "GetByName":
+                            CheckBouncyCastleEllipticCurve(containingType.Value, invocation, c);
+                            break;
+                        case "Init":
+                            CheckBouncyCastleParametersGenerators(containingType.Value, invocation, c);
+                            break;
+                        default:
+                            // Current method is not related to any cryptographic method of interest
+                            break;
                     }
+                },
+                SyntaxKind.InvocationExpression);
 
-                    var firstParameter = objectCreationSymbol.Parameters.FirstOrDefault();
-                    if (firstParameter == null)
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var objectCreation = (ObjectCreationExpressionSyntax)c.Node;
+                    if (!(c.SemanticModel.GetTypeInfo(objectCreation).Type is ITypeSymbol containingType))
                     {
                         return;
                     }
 
-                    var optionalKeyLength = c.SemanticModel.GetConstantValue(objectCreation.ArgumentList.Arguments[0].Expression);
-                    if (optionalKeyLength.HasValue &&
-                        optionalKeyLength.Value is int keyLengthValue &&
-                        keyLengthValue < MinimalKeyLengthForRsa)
+                    CheckSystemSecurityEllipticCurve(containingType, objectCreation, objectCreation.ArgumentList, c);
+                    CheckSystemSecurityCryptographyAlgorithms(containingType, objectCreation, c);
+                    CheckBouncyCastleKeyGenerationParameters(containingType, objectCreation, c);
+                },
+                SyntaxKind.ObjectCreationExpression);
+
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var assignment = (AssignmentExpressionSyntax)c.Node;
+                    var propertyName = GetPropertyName(assignment.Left);
+                    if (propertyName == "KeySize"
+                        && assignment.Left is MemberAccessExpressionSyntax memberAccess
+                        && memberAccess.Expression != null
+                        && c.SemanticModel.GetTypeInfo(memberAccess.Expression).Type is ITypeSymbol containingType)
                     {
-                        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, objectCreation.GetLocation(), MinimalKeyLengthForRsa));
+                        // Using the KeySize setter on DSACryptoServiceProvider/RSACryptoServiceProvider does not actually change the underlying key size
```

Comment:
This is serious charge. It would be nice to have link to MS docs on second line.